### ISSUE TITLE
Cloud Hypervisor uses double memory when shared memory is off with a vfio device attached

### ISSUE THREAD

**Issue by NunoDasNeves** [cloud-hypervisor/cloud-hypervisor#4805]
Running Cloud Hypervisor with a vfio device and memory config shared=off seems to use double the memory compared to shared=on, as measured by "MemAvailable" and "Active(anon)" lines from /proc/meminfo.

My theory is that the double memory is in the file system cache; due to using MAP_PRIVATE with a memfd file, it has to make a copy of the memory for the vmm process - I could be totally wrong about this... *but*:
I drafted [a patch](https://github.com/cloud-hypervisor/cloud-hypervisor/compare/main...NunoDasNeves:cloud-hypervisor:nudasnev/map_anon) which seems to fix this issue. It simply replaces usage of memfd with `MAP_ANONYMOUS` flag in `create_ram_region()`.

My results from testing different configurations with/without my patch. The guest is a 4GiB VM:
```
MemTotal: 31.07 GiB

VM size: 4G
vfio      shared    patched   MemAvailable  Active(anon)
          (baseline)          30.43 GiB      0.00 GiB
no        no        no        28.92 GiB      0.00 GiB
no        no        yes       29.67 GiB      0.44 GiB
no        yes       no        29.66 GiB      0.00 GiB
no        yes       yes       29.65 GiB      0.00 GiB
yes       no        no        22.40 GiB      4.00 GiB <- Uh oh
yes       no        yes       26.41 GiB      0.73 GiB <- MAP_ANONYMOUS seems to (mostly) fix the issue
yes       yes       no        26.40 GiB      0.39 GiB
yes       yes       yes       26.40 GiB      0.39 GiB
```
The full /proc/meminfo logs are attached below.

I tried using [vmm_memory](https://github.com/sameo/vmm_memory) to measure 'Total Overhead', but it relies on each line of /proc/pid/smaps starting with "Size:" which it does not on my OS, so it just adds up all the memory in smaps.
**The total memory in smaps does *not* show this double memory usage** - it just shows ~0.7 GiB with vfio off, and ~4 GiB with vfio on, in all the above cases.

I'm still investigating this bug, would love some feedback as to what else I can check, and if anyone else can reproduce it.

**To Reproduce**
Run CLH with a vfio device, compared the memory usage with `shared=on` vs `shared=off`
See below for the command line I used

**Version**
`cloud-hypervisor v27.0-87-gdd0d007b`

`cargo build --release`

**VM configuration**
I used the following command, testing with shared=off and shared=on, and with and without the --device argument:
`sudo cloud-hypervisor --kernel hypervisor-fw --disk path=focal-server-cloudimg-amd64.raw --cpus boot=4 --memory size=4G,shared=off --net tap=,mac=,ip=,mask= --device path=/sys/bus/pci/devices/0000:01:00.1 -v --log-file ch.log`

`/sys/bus/pci/devices/0000:01:00.1` is a device I bound to the vfio-pci driver as per [docs/vfio.md](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/vfio.md)

Guest OS version details:
ubuntu cloud image from README

Host OS version details:
Mariner with 5.10.47 kernel
(I need to test on something newer...)

**Logs**
Output of `cat /proc/meminfo` attached for all cases (proc_meminfo_*.txt)
[proc_meminfo_vfio_no_shared_no_patched_no.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9832027/proc_meminfo_vfio_no_shared_no_patched_no.txt)
[proc_meminfo_vfio_no_shared_no_patched_yes.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9832028/proc_meminfo_vfio_no_shared_no_patched_yes.txt)
[proc_meminfo_vfio_no_shared_yes_patched_no.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9832029/proc_meminfo_vfio_no_shared_yes_patched_no.txt)
[proc_meminfo_vfio_no_shared_yes_patched_yes.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9832030/proc_meminfo_vfio_no_shared_yes_patched_yes.txt)
[proc_meminfo_vfio_yes_shared_no_patched_no.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9832031/proc_meminfo_vfio_yes_shared_no_patched_no.txt)
[proc_meminfo_vfio_yes_shared_no_patched_yes.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9832032/proc_meminfo_vfio_yes_shared_no_patched_yes.txt)
[proc_meminfo_vfio_yes_shared_yes_patched_no.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9832033/proc_meminfo_vfio_yes_shared_yes_patched_no.txt)
[proc_meminfo_vfio_yes_shared_yes_patched_yes.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9832034/proc_meminfo_vfio_yes_shared_yes_patched_yes.txt)

Output of `-v` attached for all cases (ch_*.log)
[ch_vfio_no_shared_yes_patched_no.log](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831919/ch_vfio_no_shared_yes_patched_no.log)
[ch_vfio_no_shared_yes_patched_yes.log](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831921/ch_vfio_no_shared_yes_patched_yes.log)
[ch_vfio_yes_shared_no_patched_no.log](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831922/ch_vfio_yes_shared_no_patched_no.log)
[ch_vfio_yes_shared_no_patched_yes.log](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831923/ch_vfio_yes_shared_no_patched_yes.log)
[ch_vfio_yes_shared_yes_patched_no.log](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831924/ch_vfio_yes_shared_yes_patched_no.log)
[ch_vfio_yes_shared_yes_patched_yes.log](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831925/ch_vfio_yes_shared_yes_patched_yes.log)
[ch_vfio_no_shared_no_patched_no.log](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831926/ch_vfio_no_shared_no_patched_no.log)
[ch_vfio_no_shared_no_patched_yes.log](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831927/ch_vfio_no_shared_no_patched_yes.log)

Output of `vmm_memory [vmm PID] --size=[guest memory size in kiB]` attached for all cases (vmm_memory_*.txt)
[vmm_memory_vfio_no_shared_yes_patched_no.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831930/vmm_memory_vfio_no_shared_yes_patched_no.txt)
[vmm_memory_vfio_no_shared_yes_patched_yes.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831931/vmm_memory_vfio_no_shared_yes_patched_yes.txt)
[vmm_memory_vfio_yes_shared_no_patched_no.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831932/vmm_memory_vfio_yes_shared_no_patched_no.txt)
[vmm_memory_vfio_yes_shared_no_patched_yes.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831934/vmm_memory_vfio_yes_shared_no_patched_yes.txt)
[vmm_memory_vfio_yes_shared_yes_patched_no.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831935/vmm_memory_vfio_yes_shared_yes_patched_no.txt)
[vmm_memory_vfio_yes_shared_yes_patched_yes.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831936/vmm_memory_vfio_yes_shared_yes_patched_yes.txt)
[vmm_memory_vfio_no_shared_no_patched_no.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831937/vmm_memory_vfio_no_shared_no_patched_no.txt)
[vmm_memory_vfio_no_shared_no_patched_yes.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/9831938/vmm_memory_vfio_no_shared_no_patched_yes.txt)

**Comment by rbradford** [comment]
@NunoDasNeves This might be helpful: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/6e6b519a74936cc36d59706f8b3767097170e32c/tests/integration.rs#L1586

This is the code that we use for calculating the VMM "overhead" but might not relevant for your use case

**Comment by liuw** [comment]
> My theory is that the double memory is in the file system cache; due to using MAP_PRIVATE with a memfd file, it has to make a copy of the memory for the vmm process

If this is true, how are the two copies kept in sync?

**Comment by NunoDasNeves** [comment]
I looked at some more fields in /proc/meminfo. For [reference](https://www.kernel.org/doc/Documentation/filesystems/proc.txt):
```
AnonPages: Non-file backed pages mapped into userspace page tables
Active: Memory that has been used more recently and usually not reclaimed unless absolutely necessary.
Mapped: files which have been mmaped, such as libraries
Shmem: Total memory used by shared memory (shmem) and tmpfs
```
This is interesting:
```
vfio      shared    patched   MemAvailable   Active(anon)   AnonPages      Mapped         Shmem
yes       no        no        22.40 GiB      4.00 GiB       4.13 GiB       0.13 GiB       4.00 GiB
yes       no        yes       26.41 GiB      0.72 GiB       4.13 GiB       0.13 GiB       0.00 GiB
yes       yes       no        26.40 GiB      0.39 GiB       0.13 GiB       4.13 GiB       4.00 GiB
yes       yes       yes       26.40 GiB      0.39 GiB       0.13 GiB       4.13 GiB       4.00 GiB
```
Note the first line.
Keeping with my theory outlined above, my interpretation is: There's 4GiB of `AnonPages` (the copy necessitated by MAP_PRIVATE), and also 4GiB of tmpfs (memfd file), shown by `Shmem`.

I'll admit I'm not sure why the `Mapped` column looks like that -
I thought it would correlate with memfd vs `MAP_ANONYMOUS` (i.e. `patched`), but it seems to correlate with `MAP_SHARED`/`MAP_PRIVATE` (i.e. `shared`).
I assume since `MAP_PRIVATE` has to copy the file, so it's not *really* mapping the file, just a copy, so `Mapped` is ~0 GiB for those rows?

**Comment by NunoDasNeves** [comment]
> > My theory is that the double memory is in the file system cache; due to using MAP_PRIVATE with a memfd file, it has to make a copy of the memory for the vmm process
> 
> If this is true, how are the two copies kept in sync?

@liuw They are not:
```
MAP_PRIVATE
              Create a private copy-on-write mapping.  Updates to the
              mapping are not visible to other processes mapping the
              same file, and are not carried through to the underlying
              file.  It is unspecified whether changes made to the file
              after the mmap() call are visible in the mapped region.
```

But I don't really know if the copy would be in the filesystem cache or somewhere else (it doesn't show up in the `Cache` line in /proc/meminfo). See my latest comment for more.

**Comment by NunoDasNeves** [comment]
> @NunoDasNeves This might be helpful:
> 
> https://github.com/cloud-hypervisor/cloud-hypervisor/blob/6e6b519a74936cc36d59706f8b3767097170e32c/tests/integration.rs#L1586
> 
> This is the code that we use for calculating the VMM "overhead" but might not relevant for your use case

@rbradford Glancing at the code, I think this is the exact same code as the vmm_memory utility. See my original post:
> I tried using [vmm_memory](https://github.com/sameo/vmm_memory) to measure 'Total Overhead', but it relies on each line of /proc/pid/smaps starting with "Size:" which it does not on my OS, so it just adds up all the memory in smaps.
> The total memory in smaps does not show this double memory usage - it just shows ~0.7 GiB with vfio off, and ~4 GiB with vfio on, in all the above cases.

**Comment by NunoDasNeves** [comment]
My test script: https://github.com/NunoDasNeves/cloud-hypervisor/blob/nudasnev/test-mem-script/test_mem.sh

**Comment by NunoDasNeves** [comment]
I edited my previous comment to include `Cached` and some discussion points. I will stop looking at this for today...

**Comment by sboeuf** [comment]
>My theory is that the double memory is in the file system cache; due to using MAP_PRIVATE with a memfd file, it has to make a copy of the memory for the vmm process - I could be totally wrong about this... but:
I drafted [a patch](https://github.com/cloud-hypervisor/cloud-hypervisor/compare/main...NunoDasNeves:cloud-hypervisor:nudasnev/map_anon) which seems to fix this issue. It simply replaces usage of memfd with MAP_ANONYMOUS flag in create_ram_region().

When no backing file is provided (which is the case of your testing here), we always end up creating an fd from `memfd_create()`. And the doc from this syscall says:
```
       memfd_create() creates an anonymous file and returns a file
       descriptor that refers to it.  The file behaves like a regular
       file, and so can be modified, truncated, memory-mapped, and so
       on.  However, unlike a regular file, it lives in RAM and has a
       volatile backing storage.  Once all references to the file are
       dropped, it is automatically released.  Anonymous memory is used
       for all backing pages of the file.  Therefore, files created by
       memfd_create() have the same semantics as other anonymous memory
       allocations such as those allocated using mmap(2) with the
       MAP_ANONYMOUS flag.
```
Therefore I don't think you've changed anything given you simply used `MAP_ANONYMOUS` with `mmap()`, which is supposed to have the same semantics as `memfd_create()`.

**Comment by sboeuf** [comment]
Given you're seeing this behavior only when attaching a VFIO device, it must be related to the fact that VFIO pins guest pages.
Pinning the pages means they're faulted in, so that they're available and not swapped out.
And looking at the `mmap()` documentation, I can see that `MAP_POPULATE` says the following:
```
MAP_POPULATE (since Linux 2.5.46)
              Populate (prefault) page tables for a mapping.  For a file
              mapping, this causes read-ahead on the file.  This will
              help to reduce blocking on page faults later.  The mmap()
              call doesn't fail if the mapping cannot be populated (for
              example, due to limitations on the number of mapped huge
              pages when using MAP_HUGETLB).  MAP_POPULATE is supported
              for private mappings only since Linux 2.6.23.
```
So I'm wondering, if `MAP_POPULATE` isn't supported for shared mappings, this might mean prefaulting pages might behave differently with `MAP_SHARED`.

And of course something obvious is that because you start a 4GiB VM, with a VFIO device that means you end up with CH process consuming 4GiB right away given all the pages have been brought in. What I'm trying to say is that it's fully expected with VFIO to see a VMM consuming the equivalent of the guest RAM size without even running anything in the guest.
And the reason why you don't see these 4GiB with `MAP_SHARED` is simply because it behaves differently.

**Comment by NunoDasNeves** [comment]
@sboeuf To address your last point:
> And the reason why you don't see these 4GiB with `MAP_SHARED` is simply because it behaves differently. (But I think at some point the kernel is going to account for it).

We DO see **4GiB** used with `MAP_SHARED`.
For `MAP_PRIVATE`, we see **8GiB** used.
I think I did a poor job of highlighting this in my initial post; I'm sorry.

Here are the relevant stats from /proc/meminfo summarized from my initial post, comparing `MAP_PRIVATE` and `MAP_SHARED` with vfio on.
```
vfio      shared      MemAvailable 
    (baseline)     30.43 GiB  <- We have 30G available
yes       no       22.40 GiB  <- MAP_PRIVATE seems to use 8GiB
yes       yes      26.40 GiB  <- MAP_SHARED seems to use 4GiB
```

> Have you tried waiting for a while and see if eventually the pages are accounted?

I'm waiting until after the guest login prompt appears (my script waits 16 seconds); I thought that should be sufficient.

**Comment by NunoDasNeves** [comment]
@sboeuf Addressing the other points:

> When no backing file is provided (which is the case of your testing here), we always end up creating an fd from `memfd_create()`. And the doc from this syscall says:
> 
> ```
>        memfd_create() creates an anonymous file and returns a file
>        descriptor that refers to it.  The file behaves like a regular
>        file, and so can be modified, truncated, memory-mapped, and so
>        on.  However, unlike a regular file, it lives in RAM and has a
>        volatile backing storage.  Once all references to the file are
>        dropped, it is automatically released.  Anonymous memory is used
>        for all backing pages of the file.  Therefore, files created by
>        memfd_create() have the same semantics as other anonymous memory
>        allocations such as those allocated using mmap(2) with the
>        MAP_ANONYMOUS flag.
> ```
> 
> Therefore I don't think you've changed anything given you simply used `MAP_ANONYMOUS` with `mmap()`, which is supposed to have the same semantics as `memfd_create()`.

The semantics may be the same, but something different is clearly happening in the kernel, as /proc/meminfo shows:
```
vfio      shared    patched   MemAvailable   Active(anon)   AnonPages     Mapped        Shmem         Cached
yes       no        no        22.40 GiB      4.00 GiB       4.13 GiB      0.14 GiB      4.00 GiB      11.25 GiB
yes       no        yes       26.41 GiB      0.71 GiB       4.13 GiB      0.13 GiB      0.00 GiB      7.25 GiB
```
The first line is with memfd, the second line is with `MAP_ANONYMOUS`. Clearly the kernel is accounting these differently.


> Given you're seeing this behavior only when attaching a VFIO device, it must be related to the fact that VFIO pins guest pages.
> Pinning the pages means they're faulted in, so that they're available and not swapped out.

Yes, totally agree.
I was originally investigating this same issue with MSHV instead of KVM, where we pin the guest pages regardless of VFIO.
Then I discovered it also happens with KVM, but only when VFIO is enabled (clearly due to the pinning).

> And looking at the `mmap()` documentation, I can see that `MAP_POPULATE` says the following:
> 
> ```
> MAP_POPULATE (since Linux 2.5.46)
>               Populate (prefault) page tables for a mapping.  For a file
>               mapping, this causes read-ahead on the file.  This will
>               help to reduce blocking on page faults later.  The mmap()
>               call doesn't fail if the mapping cannot be populated (for
>               example, due to limitations on the number of mapped huge
>               pages when using MAP_HUGETLB).  MAP_POPULATE is supported
>               for private mappings only since Linux 2.6.23.
> ```
> 
> So I'm wondering, if `MAP_POPULATE` isn't supported for shared mappings, this might mean prefaulting pages might behave differently with `MAP_SHARED`.

This is interesting, I did not notice that!
I would expect the call to pin_user_pages() to fail in the kernel if this were not allowed, but maybe not?
It may be worth investigating, but `MAP_SHARED` + vfio is not causing the double memory usage problem I outlined; it is `MAP_PRIVATE` + vfio.

**Comment by NunoDasNeves** [comment]
I did a simple test this morning: create a 16GiB VM on my 32GiB host machine.

All were run with a vfio device attached:
1. With shared=on, **the guest booted successfully**
2. With shared=off, **the guest was killed by oomkiller before finishing boot**
3. With shared=off and my `MAP_ANONYMOUS` patch, **the guest booted successfully**

I tested this 2-3 times with each case, the results were consistent.
(Side note - Before testing I allowed tmpfs to use 100% of memory: `sudo mount -o remount,size=100% /dev/shm/`)

So, it is not just the accounting in /proc/meminfo - the linux kernel thinks the VMM is using too much memory in one case but not the others.

**Comment by sboeuf** [comment]
> I did a simple test this morning: create a 16GiB VM on my 32GiB host machine.
> 
> All were run with a vfio device attached:
> 
>     1. With shared=on, **the guest booted successfully**
> 
>     2. With shared=off, **the guest was killed by oomkiller before finishing boot**
> 
>     3. With shared=off and my `MAP_ANONYMOUS` patch, **the guest booted successfully**
> 
> 
> I tested this 2-3 times with each case, the results were consistent. (Side note - Before testing I allowed tmpfs to use 100% of memory: `sudo mount -o remount,size=100% /dev/shm/`)
> 
> So, it is not just the accounting in /proc/meminfo - the linux kernel thinks the VMM is using too much memory in one case but not the others.

That's a pretty compelling proof that something's wrong :)

Could you please investigate the implementation of `memfd_create` vs `mmap` in the kernel source code that could explain this difference in behavior? I'd like to make sure we understand very well the situation before we try to propose a patch for it.

**Comment by NunoDasNeves** [comment]
> Could you please investigate the implementation of `memfd_create` vs `mmap` in the kernel source code that could explain this difference in behavior? I'd like to make sure we understand very well the situation before we try to propose a patch for it.

Yes, happy to! It may take some time...I will update this issue

**Comment by sboeuf** [comment]
I've spent some time looking at it and I think the problem comes from the use of `MAP_PRIVATE` with a file backing the mapping. Because of the Copy on Write (CoW) behavior of `MAP_PRIVATE`, when the memory is populated, this creates a copy of every page from the initial file created with `memfd_create()`.

I've created a super simple patch (based on what you observed with yours), that blindly applies `MAP_ANONYMOUS` to both cases (`MAP_SHARED` and `MAP_PRIVATE`):
```diff
diff --git a/vmm/src/memory_manager.rs b/vmm/src/memory_manager.rs
index c1a79432..02acabe7 100644
--- a/vmm/src/memory_manager.rs
+++ b/vmm/src/memory_manager.rs
@@ -1309,9 +1309,9 @@ impl MemoryManager {
 
         let mut mmap_flags = libc::MAP_NORESERVE
             | if shared {
-                libc::MAP_SHARED
+                libc::MAP_SHARED | libc::MAP_ANONYMOUS
             } else {
-                libc::MAP_PRIVATE
+                libc::MAP_PRIVATE | libc::MAP_ANONYMOUS
             };
         if prefault {
             mmap_flags |= libc::MAP_POPULATE;
```
and I can confirm this also solves the problem. The reason why this is solving the issue is because `MAP_ANONYMOUS` takes precedence over the provided file descriptor, and basically ignores it. Therefore the mapping is performed with no backing file, and that why in that case `shared=off` (`MAP_PRIVATE`) and `shared=on` (`MAP_SHARED`) behave the same way.
But this is not a viable patch given we want a file to back the guest memory for a few use cases.

If we look at the reasons for creating a file backing the guest memory, here is the list:
- `vhost-user`
- `vfio-user`
- live upgrade (migration on the same host machine)

And if we ask ourselves if we actually need `MAP_PRIVATE`, I'm starting to think the answer is no. I don't think there's any security concern about using `MAP_SHARED` over `MAP_PRIVATE`. Cloud Hypervisor is in charge of sharing the file descriptor backing the memory with other processes, meaning it has the control of it.

To summarize, we need a file to back the memory in multiple situations, and there's no security issue related to *not* using `MAP_PRIVATE`. For these reasons, it might be easier to always perform `mmap()` with `MAP_SHARED` so that we can avoid the CoW behavior that duplicates the pages, and this would give us the opportunity to simplify the CLI (no more `shared=on|off`), meaning we could also simplify a few error handling cases where we advise the user to use "shared=on" because without it, things wouldn't work.

**Comment by liuw** [comment]
FYI https://github.com/qemu/qemu/blob/79fc2fb685f35a5e71e23629760ef4025d6aba31/util/memfd.c#L133

**Comment by NunoDasNeves** [comment]
> I've spent some time looking at it and I think the problem comes from the use of `MAP_PRIVATE` with a file backing the mapping. Because of the Copy on Write (CoW) behavior of `MAP_PRIVATE`, when the memory is populated, this creates a copy of every page from the initial file created with `memfd_create()`.

Yes, this is what I suspected.

> To summarize, we need a file to back the memory in multiple situations, and there's no security issue related to **not** using `MAP_PRIVATE`. For these reasons, it might be easier to always perform `mmap()` with `MAP_SHARED` so that we can avoid the CoW behavior that duplicates the pages, and this would give us the opportunity to simplify the CLI (no more `shared=on|off`), meaning we could also simplify a few error handling cases where we advise the user to use `shared=on` because without it, things wouldn't work.

Sounds sensible!

**Comment by liuw** [comment]
I guess the conclusion is to always use `MAP_SHARED`?

**Comment by sboeuf** [comment]
Well if there's no objection and if nobody can think of a good reason to keep `MAP_PRIVATE`, then yes let's always use `MAP_SHARED`. Let's wait for @rbradford to come back from vacation to confirm he agrees with this.

**Comment by rbradford** [comment]
Thank you everyone for diving into this. Unfortunately there are currently two reasons why we cannot always use `MAP_SHARED`:

a.) The support for backing the memory with a file from the filesystem (this was used for the prototype of "fast restoring" by @praveen-pk).  See `MemoryZoneConfig`. If we switch the default to `MAP_SHARED` then the expected CoW behaviour would be lost with writes made to the file on disk. However I think we could remove this functionality as a better way to implement fast restoring would be to use `userfaultd`
b.) KSM doesn't work with `MAP_SHARED`.

My recommendation is as follows:

1. Immediately use `MAP_ANONYMOUS` if `MAP_PRIVATE` is set. One problem we have is `shared=off,hugepages=on`. We need the file backing there. So I think we should require `MAP_SHARED` if `hugepages=on`
2. Discuss dropping `MemoryZoneConfig::file`
3. Discuss dropping `MemoryConfig::mergeable`

**Comment by sboeuf** [comment]
> Thank you everyone for diving into this. Unfortunately there are currently two reasons why we cannot always use `MAP_SHARED`:
> 
> a.) The support for backing the memory with a file from the filesystem (this was used for the prototype of "fast restoring" by @praveen-pk). See `MemoryZoneConfig`. If we switch the default to `MAP_SHARED` then the expected CoW behaviour would be lost with writes made to the file on disk. However I think we could remove this functionality as a better way to implement fast restoring would be to use `userfaultd` b.) KSM doesn't work with `MAP_SHARED`.
> 
> My recommendation is as follows:
> 
>     1. Immediately use `MAP_ANONYMOUS` if `MAP_PRIVATE` is set. One problem we have is `shared=off,hugepages=on`. We need the file backing there. So I think we should require `MAP_SHARED` if `hugepages=on`
> 
>     2. Discuss dropping `MemoryZoneConfig::file`
> 
>     3. Discuss dropping `MemoryConfig::mergeable`

I agree we should fix the issue first! And I also agree we could use this opportunity to simplify the code by removing some unused functionalities.
